### PR TITLE
Bug 1999556: annotate rendered config with OCP version

### PR DESCRIPTION
**Pull request by kikisdeliveryservice**
We need to ensure that all rendered configs were generated in the proper version of the OCP release version before allowing the MCO to report upgraded. Previously, we added checks to verify MCs were generated with the correct MCO commit/version.Hash and verify the osImageURL. However, on rare occasions (and usually just in ci) there will be an OCP release that neither includes a new MCO commit nor a new rhcos image, and because we aren't checking against the OCP version, we prematurely report upgraded before the new MCO/MCC/Rendered MCs are generated.

- added [annotation](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2918/pull-ci-openshift-machine-config-operator-master-e2e-aws-workers-rhel8/1483961095910592512/artifacts/e2e-aws-workers-rhel8/gather-extra/artifacts/machineconfigs.json) of [release version](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2918/pull-ci-openshift-machine-config-operator-master-e2e-aws-workers-rhel8/1483961095910592512/artifacts/e2e-aws-workers-rhel8/gather-extra/artifacts/clusterversion.json) to rendered configs and controller config :

>              "metadata": {
>                 "annotations": {
>                     "machineconfiguration.openshift.io/release-image-version": "4.10.0-0.ci.test-2022-01-20-003840-ci-op-mnyl122f-latest"
>                 }

### CODE REVIEW

**Comment by kikisdeliveryservice** [comment]
Upgrade succeeded but hit a bz
/test e2e-agnostic-upgrade

**Comment by kikisdeliveryservice** [comment]
mco/pr looking good, but some unrelatedblocked passing

``multus-validating-config.k8s.io: x509: certificate signed by unknown authority"```

/test e2e-agnostic-upgrade

**Comment by kikisdeliveryservice** [comment]
aws not doing great
/test e2e-aws

**Comment by kikisdeliveryservice** [comment]
`error: error creating buildah builder: Error reading blob ...invalid status code from registry 503 (Service Unavailable)` on both retests =/

/retest-required

File: pkg/operator/status.go [hunk 1]
```diff
@@ -607,6 +607,14 @@ func isMachineConfigPoolConfigurationValid(pool *mcfgv1.MachineConfigPool, versi
 	if renderedMC.Spec.OSImageURL != osURL {
 		return fmt.Errorf("osImageURL mismatch for %s in %s expected: %s got: %s", pool.GetName(), renderedMC.Name, osURL, renderedMC.Spec.OSImageURL)
 	}
+
+	// check that the rendered config matches the OCP release version for cases where there is no OSImageURL change nor no new MCO commit
+	if !inClusterBringup {
```
Comment (cgwalters) [inline pkg/operator/status.go:612]: Why did you add this conditional?  The `inClusterBringup` bit came from https://github.com/openshift/machine-config-operator/commit/b8876d946f8bb87e95ce59efb7a88c3ff19fa43e and ...I no longer recall exactly what that variable means now :cry:   We should add a documentation comment for it.

I'm not saying it's not necessary, I just want to understand why it might be.

File: pkg/operator/status.go [hunk 2]
```diff
@@ -607,6 +607,14 @@ func isMachineConfigPoolConfigurationValid(pool *mcfgv1.MachineConfigPool, versi
...
+		rv, ok := renderedMC.Annotations[ctrlcommon.ReleaseImageVersionAnnotationKey]
+		if ok && rv != releaseVersion {
```
Comment (cgwalters) [inline pkg/operator/status.go:614]: Here we'll silently pass if the annotation doesn't exist.  But, can we instead actually just require that the annotation exist?  

Not requiring it is definitely a more conservative choice.  I am just trying to think through whether it also means that fixing this bug requires a new version of the operator as the upgrade *source* and not just the upgrade target.

May be at least worth a log message emitted if it doesn't exist.

File: pkg/operator/sync.go [hunk 3]
```diff
@@ -572,6 +572,12 @@ func (optr *Operator) syncMachineConfigController(config *renderConfig) error {
 	// new controller can roll out too.
 	// https://bugzilla.redhat.com/show_bug.cgi?id=1879099
 	cc.Annotations[daemonconsts.GeneratedByVersionAnnotationKey] = version.Raw
+
+	// add ocp release version as annotation to controller config, for use when
+	// annotating rendered configs with same.
+	optrVersion, _ := optr.vStore.Get("operator")
```
Comment (cgwalters) [inline pkg/operator/sync.go:578]: One thing I had to dig so I can understand this better.  This value comes from the `RELEASE_IMAGE` environment variable.  Which is in turn set in the deployment's pod spec for the operator.

And the way *that* works is:

```
        env:
          - name: RELEASE_VERSION
            value: "0.0.1-snapshot"
```

Except, when a release image is generated, that value will be substituted with the OCP release version.

Which then actually means that every upgrade will result in the operator respawning, even though the code didn't change.  I hadn't realized that.

File: pkg/controller/common/constants.go [hunk 4]
```diff
@@ -4,6 +4,9 @@ const (
 	// GeneratedByControllerVersionAnnotationKey is used to tag the machineconfigs generated by the controller with the version of the controller.
 	GeneratedByControllerVersionAnnotationKey = "machineconfiguration.openshift.io/generated-by-controller-version"
 
+	// ReleaseImageVersionAnnotationKey is used to tag the rendered machineconfigs & controller config with the release image version.
+	ReleaseImageVersionAnnotationKey = "machineconfiguration.openshift.io/release-image-version"
```
Comment (cgwalters) [inline pkg/controller/common/constants.go:8]: One big concern I had is checking whether this would cause the opposite problem of us re-rolling all nodes for an upgrade, even if the actual machineconfig content didn't change.

Because that the MCO rolls out changes to the pool based on the unique rendered config digest, so if we included the release image version in the digest, it'd cause that problem.

But, the digest (md5) is only of `config.Spec`, not the rest.  (And it really has to work that way because of the other similar annotation keys, etc.)

You and/or others may know this, I'm just writing this down for my own sanity really :smile:

**Review by cgwalters — APPROVED** [review]
Overall, this looks sane to me and I think we could ship it as is.  Some comments below which could be considered all optionals/informational.

File: pkg/operator/sync.go (hunk 3 above)
Comment (kikisdeliveryservice) [reply pkg/operator/sync.go:578]: `Which then actually means that every upgrade will result in the operator respawning`

Yeeeep. Operator does everything it would do in a "normal" (ie release involves a mco code change).

File: pkg/controller/common/constants.go (hunk 4 above)
Comment (kikisdeliveryservice) [reply pkg/controller/common/constants.go:8]: I went thru the same thing working on this if it helps :laughing:

**Comment by cgwalters** [comment]
Previous related PRs:

- https://github.com/openshift/machine-config-operator/pull/490/commits/dd54b84a9c2e7c387a846c232b8891cfaaeeaaa6
- https://github.com/openshift/machine-config-operator/pull/2206/commits/60c5ee214b5220e5499c1b687594dcf1a16ed9e8
-

File: pkg/operator/status.go (hunk 1 above)
Comment (kikisdeliveryservice) [reply pkg/operator/status.go:612]: Double checking on this, in previous iterations I hit a race when it would not be able to pass this check when the cluster was coming up. But now I'm wondering if it was more bc I was passing in the wrong value for releaseVersion. Let me try and will report back.

File: pkg/operator/status.go (hunk 2 above)
Comment (kikisdeliveryservice) [reply pkg/operator/status.go:614]: agree on adding the log :+1:

**Comment by kikisdeliveryservice** [comment]
A couple of changes are incoming

/hold

File: pkg/operator/status.go (hunk 1 above)
Comment (kikisdeliveryservice) [reply pkg/operator/status.go]: Test runs showed successful install without this check so removed.

File: pkg/operator/status.go (hunk 2 above)
Comment (kikisdeliveryservice) [reply pkg/operator/status.go]: did some testing and ended up adding a !ok check. didn't see any issues after doing some testing, so let's do it.

**Comment by kikisdeliveryservice** [comment]
Updated PR after doing some testing to implement some of @cgwalters suggestions. This is now ready. :smile_cat: 

/hold cancel

**Comment by yuqi-zhang** [comment]
Code generally lgtm. I have a question on the BZ: for situations that would trigger the bugzilla error
`the "master" pool should be updated before the CVO reports available at the new version occurred`

The `master` pool would have to be marked as `Updating` when the MCO operator reports updated right?

Looking through CI search (https://search.ci.openshift.org/?search=pool+should+be+updated+before+the+CVO+reports+available+at+the+new+version&maxAge=336h&context=1&type=junit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job), there was only 2 hits of this in the past 2 weeks. Looking at the first link, there was a new rendered-master being generated:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.7-e2e-metal-ipi-upgrade/1483817445192896512/artifacts/e2e-metal-ipi-upgrade/gather-extra/artifacts/oc_cmds/machineconfigs

Comparing those 2 in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.7-e2e-metal-ipi-upgrade/1483817445192896512/artifacts/e2e-metal-ipi-upgrade/gather-extra/artifacts/machineconfigs.json

it seems that the upgrade only contained a diff in 
`/etc/kubernetes/manifests/keepalived.yaml`

Which seems to imply that to trigger this error, a template change would be the only diff between the two OCP release versions. Would this PR also target that scenario?

File: pkg/operator/status.go [hunk 5]
```diff
@@ -607,6 +607,17 @@ func isMachineConfigPoolConfigurationValid(pool *mcfgv1.MachineConfigPool, versi
 	if renderedMC.Spec.OSImageURL != osURL {
 		return fmt.Errorf("osImageURL mismatch for %s in %s expected: %s got: %s", pool.GetName(), renderedMC.Name, osURL, renderedMC.Spec.OSImageURL)
 	}
+
+	// check that the rendered config matches the OCP release version for cases where there is no OSImageURL change nor no new MCO commit
```
Comment (yuqi-zhang) [inline pkg/operator/status.go:611]:
```suggestion
	// check that the rendered config matches the OCP release version for cases where there is no OSImageURL change nor new MCO commit
```

Comment (kikisdeliveryservice) [reply pkg/operator/status.go]: done

**Comment by kikisdeliveryservice** [comment]
@yuqi-zhang Intention for this PR is to fix the known issue of when there is an OCP upgrade that contains no new MCO commit and no new MCO osImageURL. Have had known failures there in the past, so closing the loop on that case via this pr.

In the run that you pointed to, osimageurl is same for both releases and old rendered config and new rendered config both point to same controller version `51dc0801ed7d705820f557fcabf04eff023bf568`. So, this PR should cover that case exactly. :)

**Comment by yuqi-zhang** [comment]
Yeah ok that makes sense, so we don't report before the newest render gets generated.

/lgtm

To give it a safer chance of making it in
